### DISCUSSION TITLE
Do all modules run on all servers?

### DISCUSSION THREAD

**Discussion by EthnTuttle** [fedimint/fedimint#2532] [Q&A] [answered]
Hypothetical federation with servers A through E.

A: Raspberry Pi w/ 8GB and a 1TB SSD - runs BTC full node using raspbian linux
B: Windows gaming laptop - w/ 32GB and 2TB and 8GB GPU
C: Home server - w/ 128GB and dual 24 core CPUs and expandable RAIDED 30TB
D: Stationary MBP - 16GB and 500GB SSD
E: Old Linux desktop - 8GB and 1TB

1. Do I need all modules to run on all servers? 
2. Is it one server per Guardian?

If no for #1, why not? I imagine where a federation delegates module "lifting" to various members. Obviously, C can run basically anything except for maybe GPU loads. Only B could run GPU loads and A/E might be severly limited in what runs on it due to resource constraint.
Since Guardians are already trusting each other, I would imagine module modularity (picking where they run) would be of benefit.
Another way to phrase this: Is the computational limit of your Fedimint limited to the compute power of your weakest node?

**Comment by jkitman** [comment, marked as answer]
1. Modules need to run on all servers because they need to stay in consensus (all servers have the same DB state).
2. Yes each guardian runs a server, but only >2/3s of them need to be honest/online for consensus to proceed.

"Weaker" nodes do not limit consensus but may have trouble contributing to consensus, they may contribute less items if they are slower than the others.  We are limited by the slowest of the 2/3s of nodes however.
